File: 总题库/560.SubarraySumEqualsK.cpp
// Christiano Ronaldo
// 暴力
class Solution {
public:
    int subarraySum(vector<int>& nums, int k) {
        int res = 0;
        int numsLen = nums.size();
        for (int i = 0; i < numsLen; i++) {
            int subSum = 0;
            for (int j = i; j < numsLen; j++) {
                subSum += nums[j];
                if (subSum == k) res++;
            }
        }
        return res;
    }
};

// 前缀和
class Solution {
public:
    int subarraySum(vector<int>& nums, int k) {
        unordered_map<int, int> mp;
        mp[0] = 1;
        int count = 0, pre = 0;
        for (int i : nums) {
            pre += i;
            if (mp.find(pre - k) != mp.end()) count += mp[pre - k];
            mp[pre]++;
        }
        return count;
    }
};

File: 总题库/560.SubarraySumEqualsK.py
# 暴力
class Solution:
    def subarraySum(self, nums: List[int], k: int) -> int:
        numsLen = len(nums)
        res = 0
        for i in range(numsLen):
            subSum = 0
            for j in range(i, numsLen):
                subSum += nums[j]
                if subSum == k:
                    res += 1

        return res

# 前缀和
class Solution:
    def subarraySum(self, nums: List[int], k: int) -> int:
        preSum = {}
        pre = 0
        count = 0
        preSum[0] = 1
        for i in nums:
            pre += i
            if pre - k in preSum:
                count += preSum[pre - k]
            if pre in preSum:
                preSum[pre] += 1
            else:
                preSum[pre] = 1

        return count